### PR TITLE
Scala ecosystem SDK doc

### PR DESCRIPTION
**Pull request by leo-da**
See: #675

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: docs/source/app-dev/bindings-scala/index.rst [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+- :doc:`DAML language </daml/reference/index>`
+- :doc:`Ledger API </app-dev/grpc/index>`
+- `Akka Streams API <https://doc.akka.io/docs/akka/current/stream/index.html>`_
+- `Scala programming language <https://www.scala-lang.org>`_
+
+Getting started with SBT
+************************
+
+Setting up dependencies in ``build.sbt``:
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: could you rephrase this as an instruction?

File: docs/source/app-dev/bindings-scala/index.rst [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+.. literalinclude:: ./code-snippets/quickstart-scala/build.sbt
+   :start-after: // <doc-ref:dependencies>
+   :end-before: // </doc-ref:dependencies>
+
+There are two modules in the example below:
+
+- ``scala-codegen``
+    This modules contains generated Scala ADTs.
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: Above you said "classes" - can you pick one term or the other?

File: docs/source/app-dev/bindings-scala/index.rst [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+- ``application``
+    This is the application code that makes use of the generated Scala ADTs.
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: Is this just needed to get your quickstart example working, or will everyone need this?

File: docs/source/app-dev/bindings-scala/index.rst [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    DAML to Scala code generator. Generates Scala classes from DAML models. The generated Scala code provides a type safe way of creating contracts (create command) and exercising contract choices (exercise command).
+
+The reader must be familiar with:
+
+- :doc:`DAML language </daml/reference/index>`
+- :doc:`Ledger API </app-dev/grpc/index>`
+- `Akka Streams API <https://doc.akka.io/docs/akka/current/stream/index.html>`_
+- `Scala programming language <https://www.scala-lang.org>`_
+
+Getting started with SBT
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: I don't understand - is this describing how to set up SBT in general, or for the quickstart example specifically?

File: docs/source/app-dev/bindings-scala/index.rst [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+.. literalinclude:: ./code-snippets/quickstart-scala/build.sbt
+   :start-after: // <doc-ref:generate-scala>
+   :end-before: // </doc-ref:generate-scala>
+
+Here is the entire `build.sbt file <https://github.com/digital-asset/daml/blob/master/language-support/scala/examples/quickstart-scala/build.sbt>`_
+
+Example code
+************
+
+TODO: Explain ADTs here
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: This is missing?

File: docs/source/app-dev/bindings-scala/index.rst (hunk 2 above)
Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: ADTs (algebraic data types) are classes. But not all classes are ADTs. I will use classes. Thanks.

File: docs/source/app-dev/bindings-scala/index.rst (hunk 3 above)
Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: yep, everyone will need it. This is a recommended way of setting up an application that uses scala codegen. Two modules, one module that contains generated classes and another module for the application code.

File: docs/source/app-dev/bindings-scala/index.rst (hunk 4 above)
Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: This is SBT specific setup., there is nothing quickstart specific in here.

File: docs/source/app-dev/bindings-scala/index.rst (hunk 5 above)
Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: yep, it is still WIP... it is a darf PR :)

File: docs/source/app-dev/bindings-scala/index.rst (hunk 4 above)
Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: I am just referring to something that I know works.

[the remark this answers is not in the document]
Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: you can use it from maven too, it is just we don't provide an example how to set it up in maven.

File: docs/source/app-dev/bindings-scala/index.rst [hunk 6]
```diff
@@ -0,0 +1,143 @@
+.. Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+.. SPDX-License-Identifier: Apache-2.0
+
+Scala bindings
+##############
+
+The Scala bindings is a client implementation of the *Ledger API*
+based on `Akka Streams API <https://doc.akka.io/docs/akka/current/stream/index.html>`_.
```
Comment (S11001001) [inline docs/source/app-dev/bindings-scala/index.rst]: Since akka-streams is optional, we might not say that it is "based on" it here.

File: docs/source/app-dev/bindings-scala/index.rst [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+Introduction
+============
+
+The Scala bindings library lets you write applications that connect to the Digital Asset distributed ledger using the Scala programming language.
+
+There are two main components:
+
+- Akka Streams-based API
+    The API that you use to send commands to the ledger and receive transactions back.
```
Comment (S11001001) [inline docs/source/app-dev/bindings-scala/index.rst]: Likewise, we probably want to mention a third component, the akka-less API.

Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: @stephencompall-DA what is that 3rd component? We have not tested Scala codegen with anything else but akka-based API.

Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: you technically can send generated commands using any existing API, doesn't have to be scala akka-based bindings, can be java rx-based api too. But the 3rd component... is something I am not aware of.

File: docs/source/app-dev/bindings-scala/index.rst [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+- :doc:`DAML language </daml/reference/index>`
+- :doc:`Ledger API </app-dev/grpc/index>`
+- `Akka Streams API <https://doc.akka.io/docs/akka/current/stream/index.html>`_
+- `Scala programming language <https://www.scala-lang.org>`_
+- :ref:`assistant-manual-building-dars`
+
+Getting started with sbt
+========================
+
+Set up the following dependencies in your ``build.sbt``:
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: What build.sbt? Do you expect them to have this file already?

Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: Do you want to tell them to get the quickstart-scala example first?

File: docs/source/app-dev/bindings-scala/index.rst [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+In order to use the Scala bindings, you should be familiar with:
+
+- :doc:`DAML language </daml/reference/index>`
+- :doc:`Ledger API </app-dev/grpc/index>`
+- `Akka Streams API <https://doc.akka.io/docs/akka/current/stream/index.html>`_
+- `Scala programming language <https://www.scala-lang.org>`_
+- :ref:`assistant-manual-building-dars`
+
+Getting started with sbt
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: I'm a bit confused as to what you're doing here. Are you telling people how to do something, or are you showing them what an example build.sbt file looks like? You need to explicitly tell the reader what the purpose of the information you're giving them is.

File: docs/source/app-dev/bindings-scala/index.rst [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+   :start-after: // <doc-ref:modules>
+   :end-before: // </doc-ref:modules>
+
+``scala-codegen`` module uses the following function to generate Scala classes from a DAR file.
+
+.. literalinclude:: ./code-snippets/quickstart-scala/build.sbt
+   :start-after: // <doc-ref:generate-scala>
+   :end-before: // </doc-ref:generate-scala>
+
+Here is the entire `build.sbt file <https://github.com/digital-asset/daml/blob/master/language-support/scala/examples/quickstart-scala/build.sbt>`_
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]:
```suggestion
You can get the entire `build.sbt file <https://github.com/digital-asset/daml/blob/master/language-support/scala/examples/quickstart-scala/build.sbt>`__ from the ``daml`` repository on GitHub.
```

I'm a bit anxious about this link - if you move or rewrite the example, it'll just break.

File: docs/source/app-dev/bindings-scala/index.rst [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+Generating Scala Code from Command Line
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: Not sure why the capital letters here?

```suggestion
Generating Scala code from the command line
```

File: docs/source/app-dev/bindings-scala/index.rst [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+=======================================
+
+The above example demonstrates how to use Scala codegen from **sbt**. You can also call Scala codegen directly
+from a command line. For information on how to build a **DAR** file from a **DAML** model see :ref:`assistant-manual-building-dars`.
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: "For information on how to build a **DAR** file" comes completely out of the blue here - you haven't told anyone why they'd need to care about it at all! 

I'd suggest rewriting this as a set of instructions:

1. If you're not on the latest, upgrade the SDK
2. Download the codegen tool
3. Build the DAR file (this is the right place to link to instructions)
4. Run the codegend tool.

The reason for this is that you need to help people out: you're giving them a bunch of information but it's not in the logical order that they'd need to follow it in, so it makes it much tougher to use this doc.

File: docs/source/app-dev/bindings-scala/index.rst [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+    Number of generated templates: 3
+    Number of not generated templates: 0
+    Details:
+
+The output above tells that codegen produced Scala classes for 3 templates without errors (empty ``Details:`` line).
+
+Example code
+============
+
+This section refers to the IOU DAML example from the :doc:`Quickstart guide </getting-started/quickstart>` and
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: You haven't told users what this section is about! Can you write a brief introduction saying what they'll learn from reading it - eg, this section shows you how to use the Scala bindings to achieve common tasks?

File: docs/source/app-dev/bindings-scala/index.rst [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+`quickstart-scala sbt example <https://github.com/digital-asset/daml/tree/master/language-support/scala/examples/quickstart-scala>`_.
+
+Please keep in mind that all Scala examples compiled with ``-Xsource:2.13`` **scalac** option.
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: Can you tell me why this is important?

File: docs/source/app-dev/bindings-scala/index.rst [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+
+Create a Contract and Send a Create Command
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]:
```suggestion
Create a contract and send a create command
```

Again, no need for capital letters.

Aren't 'creating a contract' and 'sending a create command' the same thing?

File: docs/source/app-dev/bindings-scala/index.rst [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+-------------------------------------------
+
+To create a Scala class representing an ``Iou`` contract:
+::
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: You don't need the :: on a separate line, you can have it at the end of the previous one.

File: docs/source/app-dev/bindings-scala/index.rst [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+
+    import com.digitalasset.ledger.client.binding.{Primitive => P}
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: Could you `literalinclude` this from somewhere rather than having it inline? Otherwise there's no guarantee this code won't break.

File: docs/source/app-dev/bindings-scala/index.rst [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+      issuer = issuer,
+      owner = issuer,
+      currency = "USD",
+      amount = BigDecimal("1000.00"),
+      observers = List())
+
+To send a corresponding create command:
+::
+
+    val createCmd = iou.create
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: Same comment as above, can you use a literalinclude?

File: docs/source/app-dev/bindings-scala/index.rst [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+    clientUtil.submitCommand(issuer, issuerWorkflowId, createCmd)
+
+For more details, please refer to the implementation of `com.digitalasset.quickstart.iou.ClientUtil#submitCommand <https://github.com/digital-asset/daml/blob/master/language-support/scala/examples/quickstart-scala/application/src/main/scala/com/digitalasset/quickstart/iou/ClientUtil.scala>`_.
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: What exactly will they find out from reading that? ie - what are the more details?

File: docs/source/app-dev/bindings-scala/index.rst [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+
+Receive a Transaction, Exercise a Choice and Send an Exercise Command
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]:
```suggestion
Receive a transaction, exercise a choice and send an exercise command
```

Receive a transaction feels like a separate task to me?

File: docs/source/app-dev/bindings-scala/index.rst [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+---------------------------------------------------------------------
+
+To receive a transaction and decode a create event for ``IouTransfer`` contract:
+::
+
+    private val newOwner = P.Party("Bob")
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: Same comment as above - please use a literalinclude

File: docs/source/app-dev/bindings-scala/index.rst [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+      decodeCreated[M.IouTransfer](tx).foreach { contract: Contract[M.IouTransfer] =>
+        logger.info(s"$newOwner received contract: $contract")
+        ...
+      }
+    }
+
+To exercise ``IouTransfer_Accept`` choice on the received ``IourTransfer`` contract and send a corresponding exercise command:
+::
+
+    val exerciseCmd = contract.contractId.exerciseIouTransfer_Accept(actor = newOwner)
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: Same comment as above - please use a literalinclude

File: docs/source/app-dev/bindings-scala/index.rst (hunk 21 above)
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: Could you also write some comments in this code to explain what it's doing?

File: docs/source/app-dev/bindings-scala/index.rst [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+    clientUtil.submitCommand(newOwner, newOwnerWorkflowId, exerciseCmd)
+
+Fore more details, please refer to the `com.digitalasset.quickstart.iou.IouMain#newOwnerAcceptsAllTransfers <https://github.com/digital-asset/daml/blob/master/language-support/scala/examples/quickstart-scala/application/src/main/scala/com/digitalasset/quickstart/iou/IouMain.scala>`_.
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: Same as above - what will they learn from this? Be specific as to why they should follow the link, what it will help them with.

File: docs/source/app-dev/bindings-scala/index.rst [hunk 24]
```diff
@@ -0,0 +1,144 @@
+.. Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+.. SPDX-License-Identifier: Apache-2.0
+
+Scala bindings
+##############
+
+This tutorial provides a basic Scala programmer's introduction to working with Digital Asset distributed ledger,
```
Comment (bethaitman) [inline docs/source/app-dev/bindings-scala/index.rst]: I wouldn't call this a tutorial - a tutorial walks you through step by step how to achieve something. I would just say "this page".

File: docs/source/app-dev/bindings-scala/index.rst (hunk 10 above)
Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: we should not be moving it. I thought the RST doc build tool actually checks if links exist

Comment (bethaitman) [reply docs/source/app-dev/bindings-scala/index.rst]: It checks if links internal to the docs exist. It doesn't have a way of checking links to external websites!

Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: I will see if I can change it to point to the symlink inside the docs that I created

Comment (bethaitman) [reply docs/source/app-dev/bindings-scala/index.rst]: I don't think that'll work - as far as I'm aware, that code file won't be published, so you won't be able to link to it within the docs site.

You might be able to make it a download link instead? Search "sphinx download directive" for how to do that - you'll want to preview it locally and make sure it works, though

Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: In this case I will leave it for later as a tech debt (issue will be created)... I have a hard deadline to finish the docs by Thursday. We will not be moving the examples anywhere any time soon.

Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: #1134

File: docs/source/app-dev/bindings-scala/index.rst (hunk 21 above)
Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: I don't think that creating a Party instance needs explanations.

Comment (bethaitman) [reply docs/source/app-dev/bindings-scala/index.rst]: I just put the comment here as it was the start of the code snippet - I leave it to you to best decide which bits of the code are worthwhile commenting. But in general just giving some descriptive comments is helpful in documentation - more than you would in your own source code

Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: I think the doc reference comments like the logging calls should be more than enough to figure out what the following code snippet supposed to do:
``` 
    // <doc-ref:subscribe-and-decode-iou-transfer>
    _ <- clientUtil.subscribe(newOwner, offset0, None) { tx =>
      logger.info(s"$newOwner received transaction: $tx")
```
Sorry, I will not add more comments explaining what the example does. I prefer to receive a feed back from the developers looking at the example and make it cleaner if needed.

File: docs/source/app-dev/bindings-scala/index.rst [hunk 25]
```diff
@@ -0,0 +1,154 @@
...
+   :start-after: // <doc-ref:new-owner-definition>
+   :end-before: // </doc-ref:new-owner-definition>
+
+and the following code that handles subscription and decoding:
+
+.. literalinclude:: ./code-snippets/quickstart-scala/application/src/main/scala/com/digitalasset/quickstart/iou/IouMain.scala
+   :start-after: // <doc-ref:subscribe-and-decode-iou-transfer>
+   :end-before: // </doc-ref:subscribe-and-decode-iou-transfer>
+
+To exercise ``IouTransfer_Accept`` choice on the ``IourTransfer`` contract that you received and send a corresponding :ref:`com.digitalasset.ledger.api.v1.exercisecommand`:
```
Comment (S11001001) [inline docs/source/app-dev/bindings-scala/index.rst]:
```suggestion
To exercise ``IouTransfer_Accept`` choice on the ``IouTransfer`` contract that you received and send a corresponding :ref:`com.digitalasset.ledger.api.v1.ExerciseCommand`:
```

Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: thanks... but the ref link is no caps.

Comment (leo-da) [reply docs/source/app-dev/bindings-scala/index.rst]: actually the doc build tool takes references in camel notation too and converts them to all small letters. I like camel... more readable. Going to check the rest of the link in the doc.
